internal/proxy/reverseproxy/reverseproxy_test.go

```diff
@@ -0,0 +1,147 @@
+// Reverse proxy tests.
+
+package reverseproxy
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus/hooks/test"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestReverseProxy(t *testing.T) {
+	const backendResponse = "I am the backend"
+	const backendStatus = 404
+	retries := 0
+	jenkins := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		// simulate connection termination by closing the connection
+		// if the "mode" is "hangup"
+		if r.Method == "GET" && r.FormValue("mode") == "hangup" {
+			if retries < 5 {
+				// NOTE Hijack results in calling  the handlerfunc again for the
+				// first time so for the first time retry will be incremented to 2
+				retries++
+				c, _, _ := w.(http.Hijacker).Hijack()
+				// must not have any affect since the connection is going to be closed
+				defer c.Close()
+			} else {
+				w.WriteHeader(200)
+				w.Write([]byte("all good"))
+			}
+			return
+		}
+
+		assert.Len(t, r.TransferEncoding, 0, "backend got unexpected TransferEncoding")
+		assert.NotEmpty(t, r.Header.Get("X-Forwarded-For"), "didn't get X-Forwarded-For header")
+		assert.Empty(t, r.Header.Get("Connection"), "handler got Connection header value")
+		assert.Empty(t, r.Header.Get("Upgrade"), "handler got Upgrade header value")
+		assert.Empty(t, r.Header.Get("Proxy-Connection"), "handler got Proxy-Connection header value")
+
+		w.Header().Set("Trailers", "not a special header field name")
+		w.Header().Set("Trailer", "X-Trailer")
+		w.Header().Set("X-Foo", "bar")
+		w.Header().Set("Upgrade", "foo")
+		w.Header().Add("X-Multi-Value", "foo")
+		w.Header().Add("X-Multi-Value", "bar")
+		http.SetCookie(w, &http.Cookie{Name: "flavor", Value: "chocolateChip"})
+		w.WriteHeader(backendStatus)
+		w.Write([]byte(backendResponse))
+		w.Header().Set("X-Trailer", "trailer_value")
+		w.Header().Set(http.TrailerPrefix+"X-Unannounced-Trailer", "unannounced_trailer_value")
+	}))
+	defer jenkins.Close()
+
+	jenkinsURL, err := url.Parse(jenkins.URL)
+	assert.NoError(t, err, "url parse error")
+
+	// this fakes the jenkins proxy that which sets the request URL
+	fakeJenkinsProxy := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			requestURL := *r.URL
+
+			r.Host = jenkinsURL.Host
+			r.URL.Host = jenkinsURL.Host
```

Comment:
https://stackoverflow.com/questions/42921567/what-is-the-difference-between-host-and-url-host-for-golang-http-request says that `request.Host` and `request.URL.Host` are always different. Can you confirm that what you are doing is right?

Reply:
yes, I am sure about this part. The req.Host is used by openshift LB to redirect the call to actual jenkins. r.URL.Host is host the request is being sent to so both need to be the same. So that the request reaches LB which inspects req.Host to direct the request to actual jenkins.